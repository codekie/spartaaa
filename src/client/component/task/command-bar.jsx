import React, { PureComponent } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Panel from 'react-bulma-components/lib/components/panel';
import Button from 'react-bulma-components/lib/components/button';
import Icon from 'react-bulma-components/lib/components/icon';
import { faPlay, faStop, faStepForward, faStepBackward } from '@fortawesome/free-solid-svg-icons';

import './command-bar.scss';
import PropTypes from 'prop-types';

const TAG__NEXT = 'next';

export default class Task extends PureComponent {
    static propTypes = {
        activateTask: PropTypes.func.isRequired,
        deactivateTask: PropTypes.func.isRequired,
        toggleNext: PropTypes.func.isRequired,

        isActive: PropTypes.bool,
        isCompleted: PropTypes.bool,
        taskId: PropTypes.number,
        tags: PropTypes.array.isRequired
    };

    constructor(props) {
        super(props);
    }
    render() {
        const props = this.props,
            { isActive, isCompleted, activateTask, deactivateTask, taskId, tags, toggleNext } = props;
        return (
            <Panel className="task-command-bar">
                { _createStartStopButton({ taskId, isCompleted, isActive, deactivateTask, activateTask }) }
                { _createToggleNextButton({ taskId, isCompleted, tags, toggleNext }) }
            </Panel>
        );
    }
}

function _createStartStopButton({ taskId, isCompleted, isActive, deactivateTask, activateTask } = {}) {
    if (isCompleted) { return null; }
    return (
        <Button className="is-grouped btn-activation" onClick={() => (
            isActive
                ? deactivateTask(taskId)
                : activateTask(taskId)
        )}>
            <Icon className="is-small">
                <FontAwesomeIcon icon={isActive ? faStop : faPlay} className="task-icon" />
            </Icon>
        </Button>
    );
}

function _createToggleNextButton({ taskId, isCompleted, tags, toggleNext } = {}) {
    if (isCompleted) { return null; }
    const hasNext = tags.includes(TAG__NEXT);
    return (
        <Button className="is-grouped btn-toggle-next" onClick={() => toggleNext(taskId)}>
            <Icon className="is-small">
                <FontAwesomeIcon icon={hasNext ? faStepBackward : faStepForward} className="task-icon" />
            </Icon>
        </Button>
    );
}
