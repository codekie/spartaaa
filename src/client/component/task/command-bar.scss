@import '../../bulma-variables';

.task-command-bar {
  button { padding-top: 0; padding-bottom: 0; height: 1.7em; width: 2em; margin-right: 0.2rem; }
}
