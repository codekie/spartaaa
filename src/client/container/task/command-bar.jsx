// # IMPORTS

// 3rd-party
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
// App
import CommandBar from '../../component/task/command-bar.jsx';
import { extractImmutable } from '../../util';
import { createDispatchMapper } from '../../util/prop-mapper';
import PropTypes from 'prop-types';
import TaskStatus from '../../../comm/task-status';
import ActionType from '../../store/action-type';
import ActionCreator from '../../store/action-creators';

// # CONSTANTS

const MAP__DISPATCH_TO_PROPS = {
    [ActionType.activateTask]: ActionCreator[ActionType.activateTask],
    [ActionType.deactivateTask]: ActionCreator[ActionType.deactivateTask],
    [ActionType.finishTask]: ActionCreator[ActionType.finishTask],
    [ActionType.toggleNext]: ActionCreator[ActionType.toggleNext]
};


// # EXPORT PUBLIC API

export class ConnectedCommandBar extends PureComponent {
    static propTypes = {
        uuid: PropTypes.string.isRequired,

        activateTask: PropTypes.func.isRequired,
        deactivateTask: PropTypes.func.isRequired,
        toggleNext: PropTypes.func.isRequired,

        taskId: PropTypes.number,
        status: PropTypes.string,
        start: PropTypes.number,
        tags: PropTypes.array.isRequired
    };

    render() {
        const { taskId, tags, start, activateTask, deactivateTask, toggleNext, status } = this.props,
            isCompleted = status === TaskStatus.completed;
        return (<CommandBar taskId={taskId}
            isCompleted={isCompleted}
            isActive={!!start}
            activateTask={activateTask}
            deactivateTask={deactivateTask}
            tags={tags}
            toggleNext={toggleNext} />);
    }
}

export default connect(
    _mapStateToProps,
    createDispatchMapper(MAP__DISPATCH_TO_PROPS)
)(extractImmutable(ConnectedCommandBar));


function _mapStateToProps(state, ownProps) {
    const task = state.get('tasks').get('tasks').get(ownProps.uuid);
    return {
        taskId: task.get('id'),
        status: task.get('status'),
        start: task.get('start'),
        tags: task.get('tags').toJS()
    };
}
