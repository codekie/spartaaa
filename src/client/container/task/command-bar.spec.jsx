import React from 'react';
import { shallow } from 'enzyme';
import { ConnectedCommandBar } from './command-bar.jsx';

it('renders correctly', () => {
    const tree = shallow(<ConnectedCommandBar uuid="Manfred" activateTask={jest.fn()} deactivateTask={jest.fn()}
        toggleNext={jest.fn()} tags={[]}
    />).debug();
    expect(tree).toMatchSnapshot();
});
