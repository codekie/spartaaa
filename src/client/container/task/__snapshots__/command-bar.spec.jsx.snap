// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `"<Task taskId={[undefined]} isCompleted={false} isActive={false} activateTask={[Function: mockConstructor]} deactivateTask={[Function: mockConstructor]} tags={{...}} toggleNext={[Function: mockConstructor]} />"`;
